crates/core/tedge_agent/src/operation_workflows/actor.rs

```diff
@@ -317,31 +319,65 @@ impl WorkflowActor {
-                        let new_state = state.update(handlers.on_error.unwrap_or_else(|| {
-                            GenericStateUpdate::failed("sub-operation failed".to_string())
-                        }));
-                        self.publish_command_state(new_state).await?;
-                        self.publish_command_state(sub_state.clear()).await?;
+                    let sub_operation = sub_state.operation().unwrap_or_default();
+                    if sub_state.is_finished() {
+                        let new_state = if sub_state.is_successful() {
+                            log_file
+                                .log_info(&format!(
+                                    "=> {sub_operation} sub-operation is successful"
+                                ))
+                                .await;
+                            let sub_cmd_output = output_excerpt.extract_value_from(&sub_state);
+                            state
+                                .update_with_json(sub_cmd_output)
+                                .update(handlers.on_success)
+                        } else {
+                            log_file
+                                .log_info(&format!(
+                                    "=> {sub_operation} sub-operation failed: {}",
+                                    sub_state.failure_reason().unwrap_or_default()
+                                ))
+                                .await;
+                            state.update(handlers.on_error.unwrap_or_else(|| {
+                                GenericStateUpdate::failed("sub-operation failed".to_string())
+                            }))
+                        };
+                        self.publish_command_state(new_state, &mut log_file).await?;
+                        self.publish_command_state(sub_state.clear(), &mut log_file)
+                            .await?;
                     } else {
                         // Nothing specific has to be done: the current state has been persisted
                         // and will be resumed on completion of the sub-operation
                         // TODO: Register a timeout event
+                        log_file
+                            .log_info(&format!(
+                                "=> {sub_operation} sub-operation is still running"
+                            ))
+                            .await;
                     }
+                } else {
+                    log_file.log_info("=> sub-operation not yet launched").await;
                 };
 
                 Ok(())
             }
         }
     }
 
+    async fn process_command_state_update(
+        &mut self,
+        new_state: GenericCommandState,
+    ) -> Result<(), RuntimeError> {
+        if let Err(err) = self.workflows.apply_internal_update(new_state.clone()) {
+            error!("Fail to persist workflow operation state: {err}");
+        }
+        self.persist_command_board().await?;
```

Comment:
Wouldn't it be better if this `persist_command_board` step is pushed into the `apply_internal_update` and `apply_external_update` functions rather than relying on the callers, as the changes must be persisted whenever there's an update?